### PR TITLE
add type annotations

### PR DESCRIPTION
**Pull request by Okroshiashvili**
Add type annotations to all functions in all folders, except ```selection``` folder.

### CODE REVIEW

File: feature_engine/discretisation/decision_tree.py [hunk 1]
```diff
@@ -69,13 +72,13 @@ class DecisionTreeDiscretiser(BaseNumericalTransformer):
...
-        scoring="neg_mean_squared_error",
-        variables=None,
-        param_grid=None,
-        regression=True,
-        random_state=None,
-    ):
+        cv: int = 3,
+        scoring: str = "neg_mean_squared_error",
+        variables: Optional[List[str]] = None,
+        param_grid: Optional[dict] = None,
```
Comment (solegalli) [inline feature_engine/discretisation/decision_tree.py:78]: https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html
# For mappings, we need the types of both keys and values
x: Dict[str, float] = {'field': 2.0}

To determine the values, we need to look at the params that the tree regressor or classifier can take. I don't remember of the top of my head if all are numerical, or there are as well some boolean, and strings

**Review by solegalli — CHANGES_REQUESTED** [review]
Hi @Okroshiashvili 

Thank you so much for the changes. That was a lot of work. Much appreciated!

I have a few questions / suggestions:

- [ ] the param `variables` in the `init` for all the transformers in this PR takes a list of strings or a string, or None. So I think it should be `Union[str, List[str]]`. Could you change?

- [ ]  the fit() method does not have a `=>`. Is this because it returns self? I found this, not sure it helps:
https://mypy.readthedocs.io/en/stable/generics.html?highlight=return%20self#generic-methods-and-generic-self

- [ ] did you test with mypy and see that it passes without errors? If not, could you?

- [ ] are there more files that need changing? I think potentially the selection folder. When all files are ready, could you add mypy to the circleci config? For these you will need to update the following files as well:

tox.ini, an example here:
https://github.com/trainindata/testing-and-monitoring-ml-deployments/blob/master/packages/gradient_boosting_model/tox.ini#L45

test_requirements.txt, an example here:
https://github.com/trainindata/testing-and-monitoring-ml-deployments/blob/master/packages/gradient_boosting_model/test_requirements.txt#L13

we probably need a mypy.ini:
https://github.com/trainindata/testing-and-monitoring-ml-deployments/blob/master/packages/gradient_boosting_model/mypy.ini

Thanks a lot!

File: feature_engine/discretisation/arbitrary.py [hunk 2]
```diff
@@ -42,7 +45,12 @@ class ArbitraryDiscretiser(BaseNumericalTransformer):
         the interval boundaries. If False, it returns integers.
     """
 
-    def __init__(self, binning_dict, return_object=False, return_boundaries=False):
+    def __init__(
+        self,
+        binning_dict: dict,
```
[the remark this answers is not in the document]
Comment (Okroshiashvili) [reply feature_engine/discretisation/arbitrary.py:50]: update type annotations as suggested

File: feature_engine/discretisation/decision_tree.py (hunk 1 above)
Comment (Okroshiashvili) [reply feature_engine/discretisation/decision_tree.py:78]: update type annotations

**Comment by Okroshiashvili** [comment]
I update some annotations. However, we need to update tox.ini and create mypy.ini config file.

**Comment by Okroshiashvili** [comment]
@solegalli What about the selection folder? We agreed that temporarily I do not write type annotations for that folder. Is it finished and am I able to write annotations?

**Comment by solegalli** [comment]
There are a group of contributors adding 3 additional selectors to the folder. 

If you could rollout the type hint to the modules that are already in the selection folder in master, and we make the type hints test with mypy  part of the tests in circle ci, then I can ask the contributors to add type hints in their scripts as well, so they pass the tests.  From the moment we roll out type hints checks in circle ci, any new transformer would have to comply.

**Comment by Okroshiashvili** [comment]
Okay then. I will add type annotations for old transformers and then release mypy part of the tests in circleci.

**Comment by solegalli** [comment]
fyi #177 @Okroshiashvili

**Comment by Okroshiashvili** [comment]
@solegalli my apologies for being late. I'm too busy right now, having some urgent university tasks and I will not be able to contribute to Feature Engine next few weeks. I will get back ASAP.

**Comment by solegalli** [comment]
> 
> 
> @solegalli my apologies for being late. I'm too busy right now, having some urgent university tasks and I will not be able to contribute to Feature Engine next few weeks. I will get back ASAP.

Hey @Okroshiashvili no problem at all. Good luck with the work at University. When you pick this up, please do so from #177 as this PR is behind lots of changes that I introduced to overcome mypy errors.

Cheers
Sole

**Comment by Okroshiashvili** [comment]
> > @solegalli my apologies for being late. I'm too busy right now, having some urgent university tasks and I will not be able to contribute to Feature Engine next few weeks. I will get back ASAP.
> 
> Hey @Okroshiashvili no problem at all. Good luck with the work at University. When you pick this up, please do so from #177 as this PR is behind lots of changes that I introduced to overcome mypy errors.
> 
> Cheers
> Sole

@solegalli Thanks for your support

**Comment by solegalli** [comment]
finished and closed in #177

**Pull request closed without merging**